apps/emqx/integration_test/emqx_persistent_session_ds_SUITE.erl

```diff
@@ -658,9 +675,18 @@ t_session_replay_retry(_Config) ->
     %% Make `emqx_ds` believe that roughly half of the shards are unavailable.
     ok = emqx_ds_test_helpers:mock_rpc_result(
         fun(_Node, emqx_ds_replication_layer, _Function, [_DB, Shard | _]) ->
-            case erlang:phash2(Shard) rem 2 of
-                0 -> unavailable;
-                1 -> passthrough
+            %% When storing session data in DS, failing here could lead to a session being
+            %% considered new and not loaded, since we read non-atomically from the DB.
+            %% If getting the streams fail due to rpc problems, it leniently returns an
+            %% empty list of streams, and that results in a "new" session.
+            case is_restoring_session() of
+                true ->
+                    passthrough;
```

Comment:
This sounds weird, though I have to admit I haven't yet finished reviewing.

I think that read failure should not be equivalent to having no session record. Is there any good reason to do that?

On the test code: wouldn't it be simpler to match on `DB` and simulate failure only for the message DB? Current test code looks both super-hacky and extemely fragile.

Reply:
> I think that read failure should not be equivalent to having no session record. Is there any good reason to do that?

Just that I'm not sure how we should handle (recoverable) failures in this case.  :thinking:   
Sleep and retry?

> On the test code: wouldn't it be simpler to match on DB and simulate failure only for the message DB? Current test code looks both super-hacky and extemely fragile.

Sounds like a much better idea, I'll fix it.

Reply:
> Sleep and retry?

Good question. This warrants some discussion but as a safe bet I'd propagate this failure to the client, and let them decide if they want to retry.

Also wondering what would happen now if locking is unavailable in a mria transaction, probably something effectively similar?

Reply:
> Good question. This warrants some discussion but as a safe bet I'd propagate this failure to the client, and let them decide if they want to retry.

This comes from `get_streams` being [lenient with failures](https://github.com/emqx/emqx/blob/a9ce4d7645511be61094b001868380fb701da4b4/apps/emqx_durable_storage/src/emqx_ds_replication_layer.erl#L225-L231), so it could affect other portions of the code as well.

> Also wondering what would happen now if locking is unavailable in a mria transaction, probably something effectively similar?

Not sure I understand the question, but indeed not having transactions will lead to potential inconsistencies (at least in the builtin backend; FDB could avoid them, at least if deletions are not involved).

Reply:
> This comes from `get_streams` being lenient with failures

Oh I see now. TBH this is (again) feels like a hack: we probably don't want to hide failures on that level.

> Not sure I understand the question

I mean, I'm trying to understand what would happen to the client which is trying to open an existing session when current storage (mria) is unavailable.

Reply:
> I mean, I'm trying to understand what would happen to the client which is trying to open an existing session when current storage (mria) is unavailable.

Good question, not sure on the exact answer.  Since it uses dirty reads and writes, my guess is that it'll work with the local copy of the data (maybe hang in replicants while they can't RPC a core). :thinking:

Reply:
Changed the test code to match on the DB.